_posts/2021-02-04-behind-the-scenes-devise.markdown

```diff
@@ -0,0 +1,160 @@
+---
+layout: post
+title: "Behind The Scenes: Devise"
+date: 2021-02-04 14:00:00
+categories: ["learning", "devise"]
+author: arieljuod
+---
+
+[**Devise**](https://github.com/heartcombo/devise) is a well known solution for authentication in **Rails** applications. It's full featured (it not only adds authentication but also password recovery, email changing, session timeout, locking, ip tracking, etc.) and can be expanded to add even more (like JWT authentication).
+
+In this post, I'll go over the code related to the basic database authentication process, how it relates to [**Warden**](https://github.com/wardencommunity/warden) and some of the magic behind it. If you don't know what Warden is, I will be explaining the roll it plays for Devise in this article.
+
+<!--more-->
+
+## Gem Initialization Process
+
+Like most gems, when loaded, the entry point is `lib/devise.rb` ([link](https://github.com/heartcombo/devise/blob/master/lib/devise.rb)). Inside this file we can find a lot of code:
+- it requires many modules (like `rails`, `activesupport`, etc)
+- then it defines the Devise module which does many things:
+  - autoloads many modules for Controllers, Mailers, Tests, Strategies (we'll get back to this later)
+  - defines all the module's attribute accessors for configuration
+  - defines methods for extending devise
+  - and more
+- at the end it requires `warden` and Devise's extensions for `models`, `rails`, etc
+
+If we take a look the the files in the last part of the `lib/devise.rb` file, we can better understand how it works.
```

Comment:
```suggestion
If we take a look the files in the last part of the `lib/devise.rb` file, we can better understand how it works.
```

Reply:
sorry, that was a typo, it was meant to be `at the` not `the the`